packager/src/main/java/io/cdap/hub/publisher/GCSPublisher.java

```diff
@@ -123,13 +127,21 @@ private void publishPackage(Package pkg) throws Exception {
     }
   }
 
+  private Path resolveKeyPrefixForPackage(Package pkg) {
+    return Paths.get(
+        Optional.ofNullable(this.prefix).orElse(""),
+        "packages", pkg.getName(), pkg.getVersion());
+  }
+
   private void publishCategory(CategoryMeta categoryMeta) throws Exception {
-    String keyPrefix = String.format("%s/categories/%s/", prefix, categoryMeta.getName());
-    putFilesIfChanged(keyPrefix, categoryMeta.getIcon());
+    putFilesIfChanged(
+        Paths.get(this.prefix, "categories", categoryMeta.getName()),
+        categoryMeta.getIcon()
+    );
   }
 
   // if the specified file has changed, put it plus all extra files on gcs bucket.
-  private void putFilesIfChanged(String keyPrefix, @Nullable File file, File... extraFiles) throws IOException {
+  private void putFilesIfChanged(Path keyPrefix, @Nullable File file, File... extraFiles) throws IOException {
```

Comment:
I'm just nitpicking, but we could change the signature here to:
```suggestion
  private void putFilesIfChanged(Path keyPrefix, @Nullable File... files) throws IOException {
```